File: scripts/extract_products.py
import os
import json
import time
import requests

BASE_URL = "https://www.lakshmistores.com"
OUTPUT_DIR = "./lakshmi_store_data"
IMAGES_DIR = os.path.join(OUTPUT_DIR, "images")

os.makedirs(IMAGES_DIR, exist_ok=True)

headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
}

all_products = []
page = 1
limit = 250  # Maximum products allowed per page by Shopify

print(f"Starting export from {BASE_URL}...")

while True:
    url = f"{BASE_URL}/products.json?limit={limit}&page={page}"
    try:
        response = requests.get(url, headers=headers, timeout=15)
    except Exception as err:
        print(f"Network error on page {page}: {err}")
        break

    if response.status_code != 200:
        print(f"Failed to fetch page {page}. Status: {response.status_code}")
        break

    data = response.json()
    products = data.get("products", [])

    if not products:
        print("No more products found.")
        break

    print(f"Processing Page {page} ({len(products)} products)...")

    for item in products:
        # Extract variants (weights, packs, prices)
        variants_data = []
        for v in item.get("variants", []):
            variants_data.append({
                "id": v.get("id"),
                "title": v.get("title"),  # e.g., "1 kg", "500 g"
                "price": float(v.get("price", 0)),
                "compare_at_price": float(v.get("compare_at_price")) if v.get("compare_at_price") else None,
                "sku": v.get("sku"),
                "available": v.get("available")
            })

        # Gather image URLs
        images_list = [img.get("src") for img in item.get("images", []) if img.get("src")]
        main_image_url = images_list[0] if images_list else None
        local_image_filename = None

        # Download primary product image locally
        if main_image_url:
            try:
                clean_url = main_image_url.split("?")[0]
                ext = clean_url.split(".")[-1] if "." in clean_url else "jpg"
                if ext.lower() not in ["jpg", "jpeg", "png", "webp"]:
                    ext = "jpg"
                handle = item.get("handle") or str(item.get("id"))
                local_image_filename = f"{handle}.{ext}"
                local_path = os.path.join(IMAGES_DIR, local_image_filename)

                if not os.path.exists(local_path):
                    img_data = requests.get(main_image_url, headers=headers, timeout=10).content
                    with open(local_path, "wb") as f:
                        f.write(img_data)
            except Exception as e:
                print(f"Could not download image for {item.get('title')}: {e}")

        # Structure normalized product object
        product_record = {
            "id": item.get("id"),
            "title": item.get("title"),
            "handle": item.get("handle"),
            "product_type": item.get("product_type"),
            "vendor": item.get("vendor"),
            "tags": item.get("tags"),
            "description": item.get("body_html"),
            "price": variants_data[0]["price"] if variants_data else 0,
            "compare_at_price": variants_data[0]["compare_at_price"] if variants_data else None,
            "variants": variants_data,
            "primary_image_url": main_image_url,
            "local_image_path": f"/images/{local_image_filename}" if local_image_filename else None,
            "all_images": images_list
        }

        all_products.append(product_record)

    page += 1
    time.sleep(1)  # Polite crawl delay

# Save final structured JSON file
json_output_path = os.path.join(OUTPUT_DIR, "products.json")
with open(json_output_path, "w", encoding="utf-8") as f:
    json.dump(all_products, f, indent=2, ensure_ascii=False)

print(f"\nDone! Extracted {len(all_products)} products.")
print(f"- Data: {json_output_path}")
print(f"- Images: {IMAGES_DIR}")

File: scripts/extract-products.mjs
import fs from 'fs';
import path from 'path';
import { pipeline } from 'stream/promises';

const BASE_URL = "https://www.lakshmistores.com";
const OUTPUT_DIR = path.resolve(process.cwd(), "lakshmi_store_data");
const IMAGES_DIR = path.join(OUTPUT_DIR, "images");
const DATA_DIR = path.resolve(process.cwd(), "data");

// Create output directories
fs.mkdirSync(IMAGES_DIR, { recursive: true });

const headers = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36"
};

const allProducts = [];
let page = 1;
const limit = 250; // Maximum products per page allowed by Shopify

// Set max pages via CLI argument if desired: e.g. node extract-products.mjs --max-pages=2
const maxPagesArg = process.argv.find(arg => arg.startsWith('--max-pages='));
const maxPages = maxPagesArg ? parseInt(maxPagesArg.split('=')[1], 10) : Infinity;

// Option to download images locally: --download-images (default false to keep extraction fast)
const downloadImages = process.argv.includes('--download-images');

console.log(`🚀 Starting product extraction from ${BASE_URL}...`);
if (downloadImages) {
  console.log(`📸 Image downloading enabled: saving images to ${IMAGES_DIR}`);
} else {
  console.log(`⚡ Fast mode: referencing direct CDN image URLs (pass --download-images to save files locally).`);
}

async function downloadImageFile(url, destPath) {
  try {
    if (fs.existsSync(destPath)) return;
    const res = await fetch(url, { headers });
    if (!res.ok) return;
    const arrayBuffer = await res.arrayBuffer();
    fs.writeFileSync(destPath, Buffer.from(arrayBuffer));
  } catch (err) {
    // Non-fatal if an image fails to download
  }
}

async function main() {
  while (page <= maxPages) {
    const url = `${BASE_URL}/products.json?limit=${limit}&page=${page}`;
    console.log(`📡 Fetching page ${page}...`);

    let response;
    try {
      response = await fetch(url, { headers });
    } catch (error) {
      console.error(`❌ Network error fetching page ${page}:`, error.message);
      break;
    }

    if (!response.ok) {
      console.log(`⚠️ Completed or stopped on page ${page}. HTTP status: ${response.status}`);
      break;
    }

    const data = await response.json();
    const products = data.products || [];

    if (products.length === 0) {
      console.log("🏁 No more products found.");
      break;
    }

    console.log(`📦 Processing page ${page} (${products.length} products)...`);

    for (const item of products) {
      const variantsData = (item.variants || []).map(v => ({
        id: v.id,
        title: v.title,
        price: parseFloat(v.price) || 0,
        compare_at_price: v.compare_at_price ? parseFloat(v.compare_at_price) : null,
        sku: v.sku,
        available: Boolean(v.available)
      }));

      const imagesList = (item.images || []).map(img => img.src).filter(Boolean);
      const mainImageUrl = imagesList[0] || null;
      let localImageFilename = null;

      if (mainImageUrl) {
        const cleanUrl = mainImageUrl.split('?')[0];
        const ext = cleanUrl.split('.').pop() || 'jpg';
        const safeExt = ['jpg', 'jpeg', 'png', 'webp'].includes(ext.toLowerCase()) ? ext.toLowerCase() : 'jpg';
        const handle = item.handle || String(item.id);
        localImageFilename = `${handle}.${safeExt}`;

        if (downloadImages) {
          const localPath = path.join(IMAGES_DIR, localImageFilename);
          await downloadImageFile(mainImageUrl, localPath);
        }
      }

      const productRecord = {
        id: String(item.id),
        title: item.title,
        handle: item.handle,
        product_type: item.product_type,
        vendor: item.vendor,
        tags: item.tags,
        description: item.body_html,
        price: variantsData[0]?.price ?? 0,
        compare_at_price: variantsData[0]?.compare_at_price ?? null,
        variants: variantsData,
        primary_image_url: mainImageUrl,
        local_image_path: localImageFilename ? `/images/${localImageFilename}` : null,
        all_images: imagesList
      };

      allProducts.push(productRecord);
    }

    page += 1;
    // Delay 1 second to avoid rate-limiting
    await new Promise(resolve => setTimeout(resolve, 1000));
  }

  // Save to lakshmi_store_data/products.json
  const jsonOutputPath = path.join(OUTPUT_DIR, "products.json");
  fs.writeFileSync(jsonOutputPath, JSON.stringify(allProducts, null, 2), "utf-8");

  // Also save a copy directly to data/extracted_products.json for immediate Next.js app imports
  const dataOutputPath = path.join(DATA_DIR, "extracted_products.json");
  fs.writeFileSync(dataOutputPath, JSON.stringify(allProducts, null, 2), "utf-8");

  console.log(`\n✨ Success! Extracted ${allProducts.length} total products.`);
  console.log(`📁 Data saved to:`);
  console.log(`   - ${jsonOutputPath}`);
  console.log(`   - ${dataOutputPath}`);
}

main().catch(err => {
  console.error("Fatal extraction error:", err);
  process.exit(1);
});

